### PR TITLE
Generic filters

### PR DESCRIPTION
**Pull request by aperrin66**
Upgrade the filters by enabling Django style [lookups](https://docs.djangoproject.com/en/3.1/topics/db/queries/#field-lookups).

For example, it is now possible to use the following request to find datasets from the VIIRS instrument:
`GET <api_root>/datasets/?source__instrument__short_name=VIIRS`

See the modified [README](https://github.com/nansencenter/django-geo-spaas-rest-api/blob/35a604e20a576b4451f1a6bb3382ba76443c9055/README.md) for more details.

The previous version of the API is kept in the `v1` folder for now. It is not supposed to be modified, and won't be included in the CI unit tests. If needed, tests on the v1 can be run locally by using the `v1/runtest.py` script.

### CODE REVIEW

**Review by aanersc — COMMENTED** [review]
you define that `StrictFilterSet`class and use it in order to receive error message `Invalid lookups found:` in your API responses. DO you use it in the downstream in some other place? or just for unifying and standardizing the error output of API you do that?

File: geospaas_rest_api/filters.py [hunk 1]
```diff
@@ -1,111 +1,256 @@
...
+        model = geospaas.catalog.models.Dataset
+        fields = {
+            'id': '__all__',
+            'entry_id': '__all__',
+            'entry_title': '__all__',
+            'summary': '__all__',
+            'time_coverage_start': '__all__',
+            'time_coverage_end': '__all__',
+            'access_constraints': '__all__'
+        }
```
Comment (aanersc) [inline geospaas_rest_api/filters.py:226]:
```suggestion
        fields = '__all__'
```
just like serializers,can you write it here? if yes, then it simpler.

Comment (aanersc) [reply geospaas_rest_api/filters.py:226]: if yes, should be changed for all other ones

File: geospaas_rest_api/tests/test_read_only_api.py [hunk 2]
```diff
@@ -165,151 +157,177 @@ class DatasetFilteringTests(django.test.TestCase):
...
+        is submitted
+        """
+        response = self.client.get('/api/datasets/?sourc=1')
+        self.assertEqual(response.status_code, 400)
+        self.assertJSONEqual(response.content, [
+            "Invalid lookups found: ['sourc']"
+        ])
+
+
+class DatasetURIFilterFilteringTests(django.test.TestCase):
```
Comment (aanersc) [inline geospaas_rest_api/tests/test_read_only_api.py:317]: `FilterFiltering`? please simplify the name

File: geospaas_rest_api/filters.py (hunk 1 above)
Comment (aperrin66) [reply geospaas_rest_api/filters.py:226]: No, it does not work like that.

File: geospaas_rest_api/tests/test_read_only_api.py (hunk 2 above)
Comment (aperrin66) [reply geospaas_rest_api/tests/test_read_only_api.py]: done

**Comment by aperrin66** [comment]
> you define that `StrictFilterSet`class and use it in order to receive error message `Invalid lookups found:` in your API responses. DO you use it in the downstream in some other place? or just for unifying and standardizing the error output of API you do that?

Without this `StrictFilterSet`class, there is no error when there an invalid lookup in the request. Instead, the whole list of objects is returned unfiltered, which can be very confusing.
As for the usage of this class, all filtersets defined in filters.py inherit from it.

**Review by aanersc — DISMISSED** [review]

**Comment by azamifard** [comment]
Previousely, you have mentioned that `StrictFilterSet` class is needed. But now you have removed that `StrictFilterSet` class! What would happen if a bad api request comes in? unfiltered objects are returned in the reply?

**Comment by aperrin66** [comment]
> Previously, you have mentioned that `StrictFilterSet` class is needed. But now you have removed that `StrictFilterSet` class! What would happen if a bad api request comes in? unfiltered objects are returned in the reply?

It is not strictly needed, it's just less confusing for the user to receive an error instead of unfiltered data.
Unfortunately, the `StrictFilterSet` interferes with query arguments which are not lookups, so it is not usable.

**Review by azamifard — APPROVED** [review]

**Review by akorosov — APPROVED** [review]

**Pull request merged**